src/internal/clusterstate/v2.6.0/pfsdb.go

```diff
@@ -111,59 +61,107 @@ func validateExistingDAGs(cis []*v2_5_0.CommitInfo) error {
+		c := &commit{info: ci}
+		if ci.Origin.Kind != 4 {
+			c.id = id
+			id++
+		}
+		commits[oldCommitKey(ci.Commit)] = c
+	}
+	getCommitId := func(c *pfs.Commit) int {
+		return commits[oldCommitKey(c)].id
+	}
+	getCommitInfo := func(c *pfs.Commit) *v2_5_0.CommitInfo {
+		return commits[oldCommitKey(c)].info
+	}
+	var deleteCommits []*v2_5_0.CommitInfo
+	// Insert the commits into pfs.commits.
+	if err := func() (retErr error) {
+		ctx, end := log.SpanContext(ctx, "insertCommits")
+		defer end(log.Errorp(&retErr))
+		batcher := NewPostgresBatcher(ctx, tx, maxStmts)
+		for _, ci := range cis {
+			// Handle alias commit.
+			if ci.Origin.Kind == 4 {
+				deleteCommits = append(deleteCommits, ci)
 				continue
 			}
-			if err := addCommitProvenance(ctx, tx, ci.Commit, b.NewCommit(ci.Commit.Id)); err != nil {
-				return errors.Wrapf(err, "add commit provenance from %q to %q",
-					oldCommitKey(ci.Commit),
-					oldCommitKey(b.NewCommit(ci.Commit.Id)))
+			stmt := fmt.Sprintf(`INSERT INTO pfs.commits (commit_id, commit_set_id) VALUES ('%v', '%v')`, oldCommitKey(ci.Commit), ci.Commit.Id)
+			if err := batcher.Add(stmt); err != nil {
+				return err
 			}
 		}
-		// collect all the ALIAS commits to be deleted
-		// before 2.6, pfs.OriginKind_ALIAS = 4
-		if ci.Origin.Kind == 4 {
-			deleteCommits[oldCommitKey(ci.Commit)] = proto.Clone(ci).(*v2_5_0.CommitInfo)
+		return batcher.Close()
+	}(); err != nil {
+		return err
+	}
+	realAncestorCommits := make(map[string]*pfs.Commit)
+	// Update the ancestry and provenance of the commits.
+	if err := func() (retErr error) {
+		ctx, end := log.SpanContext(ctx, "updateCommits")
+		defer end(log.Errorp(&retErr))
+		batcher := NewPostgresBatcher(ctx, tx, maxStmts)
+		for _, ci := range cis {
+			// Skip alias commit.
+			if ci.Origin.Kind == 4 {
+				continue
+			}
+			ci := proto.Clone(ci).(*v2_5_0.CommitInfo)
+			// Update the parent and child commits.
+			if ci.ParentCommit != nil {
+				ci.ParentCommit = getRealAncestorCommit(getCommitInfo, realAncestorCommits, ci.ParentCommit)
+			}
+			var childCommits []*pfs.Commit
+			for _, child := range ci.ChildCommits {
+				childCommits = append(childCommits, getRealDescendantCommits(getCommitInfo, child)...)
+			}
+			ci.ChildCommits = childCommits
+			// Update the direct provenance.
+			var directProvenance []*pfs.Commit
+			for _, b := range ci.DirectProvenance {
+				toC := b.NewCommit(ci.Commit.Id)
+				// If the provenant commit's repo was deleted, its
+				// reference may still exist in the old provenance, so
+				// we skip mapping it over to the new model.
+				if _, ok := commits[oldCommitKey(toC)]; !ok {
+					continue
+				}
+				toC = getRealAncestorCommit(getCommitInfo, realAncestorCommits, toC)
+				directProvenance = append(directProvenance, toC)
+				from := strconv.Itoa(getCommitId(ci.Commit))
+				to := strconv.Itoa(getCommitId(toC))
+				stmt := fmt.Sprintf(`INSERT INTO pfs.commit_provenance(from_id, to_id) VALUES (%v, %v)`, from, to)
```

Comment:
would this sometimes execute identical statements requiring us to include NO CONFLICT? I'm thinking of a scenario where this loop runs on an ALIAS commit which results in `toC` being a realAncestor. Later, this will run again for `ci` where it's the real ancestor that was previously run against

Reply:
resolved - not an issue: `ci` is only set to REAL commits, and `toC` will not have duplicates because a commit cannot have provenance on multiple commits from a repo